src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java

```diff
@@ -175,4 +178,48 @@ private String adjustHexRangeBigInteger(String hexValue, String format, boolean
             return String.format(format, new BigInteger(hexValue, 16).subtract(BigInteger.ONE));
     }
     
+    /**
+     * This is used to determine if we have a valid geo hash (tier + position).
+     * 
+     * @param value
+     * @return true if valid
+     */
+    public boolean validHash(String value) {
+        try {
+            short tier = getTier(value);
+            if (validTier(tier) && validLength(tier, value)) {
+                return validPosition(tier, getPosition(value));
+            }
+        } catch (NumberFormatException e) {
+            // not a valid hex string in the first place
+        }
+        return false;
+    }
+    
+    public short getTier(String value) {
+        return Short.parseShort(value.substring(0, 2), 16);
+    }
+    
+    public long getPosition(String value) {
+        if (value.length() == 2) {
+            return 0;
+        }
+        return Long.parseLong(value.substring(2), 16);
+    }
+    
+    public boolean validTier(short tier) {
+        return tier >= 0 && tier <= 0x1f;
+    }
+    
+    public boolean validLength(short tier, String value) {
+        // determine the length of the position in hex characters
+        long posLen = Math.round(Math.ceil((double) tier / 4)) * 2;
+        // length is the tier length plus the position length
+        return value.length() == (2 + posLen);
+    }
+    
+    public boolean validPosition(short tier, long value) {
+        return value >= 0 && value < Math.round(Math.pow(2.0d, 2.0d * tier));
```

Comment:
Why are you rounding?

Reply:
Takes care of the double conversions and ensures I am getting a long

Reply:
changed to a cast

Reply:
changed to use bit shifts which is more efficient and guaranteed to be exact